File: 120-triangle/120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        def recurse(r,c):

            # base
            if r == len(grid)-1:
                return grid[r][c]
            if c > len(grid[c])-1:
                return float('inf')

            # memo cond
            if (r,c) in memo:
                return memo[(r,c)]

            # recursive - O(1) is the cost of the recursive call.- isolated
            result = grid[r][c] + min(recurse(r+1, c), recurse(r+1, c+1))
            memo[(r,c)] = result
            return result

        # main 
        memo = {}
        grid = triangle
        return recurse(0,0)

File: 120-triangle/120-triangle.cpp
class Solution {
public:
    int n,dp[202][202];
    int solve(vector<vector<int>>& t,int i,int j){
        if(i==n-1) return t[i][j];
        if(dp[i][j]!=-1) return dp[i][j];
        return dp[i][j]= t[i][j] + min(solve(t,i+1,j),solve(t,i+1,j+1));
    }
    
    int minimumTotal(vector<vector<int>>& triangle) {
        n = triangle.size();
        memset(dp, -1, sizeof dp);
        return solve(triangle,0,0);
    }
};